android/modules/calendar/src/java/ti/modules/titanium/calendar/RecurrenceRuleProxy.java

```diff
@@ -0,0 +1,376 @@
+					finalRRule.append(";");
+					break;
+				// Case for monthly recurring events.
+				case MONTHLY:
+					StringBuilder monthlyReccurencesString = new StringBuilder("BYDAY=");
+					// daysOfTheWeek dictionary is with highest priority.
+					if (this.daysOfTheWeek.length > 0) {
+						monthlyReccurencesString.append(this.daysOfTheWeek[0].getInt(this.weekNumberKey));
+						// Potentially add week start (Sunday or Monday) different from the default one.
+						monthlyReccurencesString.append(this.weekdaysMap.keySet().toArray()[this.daysOfTheWeek[0].getInt(this.weekNumberKey)]);
+					} else {
+						// Case in which we do not have items in daysOfTheWeek array.
+						monthlyReccurencesString.append(String.valueOf(this.daysOfTheMonth));
+					}
+					finalRRule.append(monthlyReccurencesString);
+					finalRRule.append(";");
+					break;
+			} // end of switch
+ 		}
+		// Handle interval.
+		if (this.interval != null) {
+			finalRRule.append("INTERVAL=" + String.valueOf(this.interval));
+			finalRRule.append(";");
+		}
+		// Handle end.
+		if (this.endDictionary.containsKey(this.until)) {
+			// Dictionary can contain only one of the keys,
+			// so what's left is a specific date end rule.
+			finalRRule.append("UNTIL=" + String.valueOf(this.endDictionary.get(this.until)));
+			finalRRule.append(";");
+		} else if (this.endDictionary.containsKey(this.count)) {
+				// End rule is with occurrence count.
+				finalRRule.append("COUNT=" + String.valueOf(this.endDictionary.get(this.count)));
+				finalRRule.append(";");
+			}
+
+		return finalRRule.toString();
+	}
+
+	// Constructor for getting a recurrence rule from the Calendar.
+	public RecurrenceRuleProxy(String nativeRRule, int calendarID, Date begin) {
+		// Only create a proxy if we have a valid recurrence rule.
+		if (nativeRRule != null && !nativeRRule.equals("")) {
+			this.eventBegin = begin;
+			parseRecurrenceRule(nativeRRule, calendarID);
+		}
+		calculateEnd();
+		calculateInterval();
+		fillFrequencyFields();
+	}
+
+	//region Methods for converting native to Kroll values
+	private void fillFrequencyFields() {
+		//reused variables in different cases
+		String days;
+		String byDay;
+		switch (this.frequency) {
+			case YEARLY:
+				Calendar cal = Calendar.getInstance();
+				cal.setTime(this.eventBegin);
+				//weeksOfTheYear
+				this.weeksOfTheYear = new int[]{cal.get(Calendar.WEEK_OF_YEAR)};
+				//monthsOfTheYear
+				this.monthsOfTheYear = new int[]{cal.get(Calendar.MONTH)};
+				days = matchExpression(".*(BYYEARDAY=[0-9]*).*", 10);
+				if (days != null) {
+					//daysOfTheYear
+					this.daysOfTheYear = new int[]{ Integer.valueOf(days)};
+				}
+				break;
+			case MONTHLY:
+				//daysOfTheMonth
+				days = matchExpression(".*(BYMONTHDAY=[0-9]*).*", 11);
+				if (days != null) {
+					this.daysOfTheMonth = new int[]{Integer.valueOf(days)};
+				}
+				//daysOfTheWeek
+				byDay = matchExpression(".*(BYDAY=[,0-9A-Z]*).*", 6);
```

Comment:
**Question:**
Do these regular expressions extract negative days?
We document that negative values are supported.

For example, a frequency type of MONTHLY with a -1 would reference the last day of the month, which is a valid use-case.

Reply:
The regular expressions are used to convert recurrence rule strings from Android format to Titanium format. I don't think we can get one with a negative value.
But I will add the support for negative values the other way around - Titanium to Android.